### ISSUE TITLE
[v1.0.1.8-FT-013] Many new errors, savegame load hangs at blank screen

### ISSUE THREAD

**Issue by Maeyanie** [AthenaSulisMinerva/CombatExtendedFastTrack#268]
Operating System: Windows 10 64-bit
RimWorld Version: 1.0.2408 rev747
CE FastTrack Version: While changing from FT-12 to FT-13

With FT-13, my savegame (which works under FT-12) hangs on load at a [blank screen](https://user-images.githubusercontent.com/206142/73148686-9d2faa80-408b-11ea-96d5-4efe68a2475d.png). The logfile contains many errors which aren't there under FT-12, coming from a wide variety of different mods.
Diff of the two logs attached, showing which errors are new... these logs were recorded one after another, and literally the only difference between them is the FT assembly.

In the diff, the "-" side is using the FT-12 binary (with FT-13 data files), the "+" side is the current (FT-13) Steam version. (I did check, the same thing happens with the Github-downloaded FT-13.)

[log12.diff](https://github.com/AthenaSulisMinerva/CombatExtendedFastTrack/files/4115504/log12.diff.txt)

**Comment by AthenaSulisMinerva** [comment]
I'm seeing lots of references to Vanilla Furniture Expanded - Security CE.

This mod is currently not explicitly supported by CE:FT, and I heard that this was still a WIP.

I'm not a C# coder, but if I may hazard a guess, we recompiled our CE DLL as part of the update to a new version number, while VFE Security's own DLL was looking for an older version.

@sumghai can you please check?

**Comment by Maeyanie** [comment]
There are also new exceptions with Pawn Rules, TD Enhancement Pack, Common Sense, Alien Races 2.0, LWM Deep Storage, and Simple Sidearms.
Aside from the last one don't even interact with CE:FT's systems directly, other than as storeable things.

**Comment by Maeyanie** [comment]
Seems to be not directly CE:FT's fault, today I'm getting the same errors still using the same local copy of FT-12 as yesterday.
Why replacing only the two CE binaries made the difference yesterday (literally the only change needed between "errors" and "fine") but not today I have no clue. I'll keep trying to diagnose things and may re-open this if I find something else.

**Comment by Maeyanie** [comment]
Ok, yes, narrowed it down and it does seem to be exactly that; something (probably VFE Security CE) is trying to load specifically Combat Extended DLL version 1.0.1.6, where the FT-13 one is 1.0.1.8.
Once that fails, everything falls apart and all sorts of mods using Harmony patches give errors, even if they don't touch CE itself.
I'm not sure if you consider this to be a problem on your end or not. Normally I'd assume not, since it just needs the other end to update, but your features list includes "Fully compatible with existing CE savegames - delete CE completely, drop in CE:FT and go!" and "Recognizes existing third-party standalone patches intended for the original CE" which this breaks, so maybe you do.

**Comment by sumghai** [comment]
I'll take it from here, @AthenaSulisMinerva.

> I'm not sure if you consider this to be a problem on your end or not. Normally I'd assume not, since it just needs the other end to update, but your features list includes "Fully compatible with existing CE savegames - delete CE completely, drop in CE:FT and go!" and "Recognizes existing third-party standalone patches intended for the original CE" which this breaks, so maybe you do.

By these statements, we mean that we recognize _**XML**_ patches written for the original upstream CE mod, which is what the vast majority of patches out there are. If a mod such as VEFS-CE wants to do unusual things with Harmony patching of C# assemblies, that's outside our guarantee.

(This is the same stance taken by the original CE team, back when the baseline CE 1.6 update changed apparel layer defs and made the third-party GraphicApparelDetour incompatible - [every change has the potential to break somebody's workflow](https://xkcd.com/1172/))

That being said, @Maeyanie, I want you to try two things:

1. Load your modlist with the following:
   * Core
   * CE:FT (latest FT-013 release, freshly downloaded from GitHub and unmodified)
   * VEFS
   * VEFS-CE
2. Start a new savegame and confirm that the baseline setup causes save loading to fail
3. Replace the DLL in `CombatExtendedFastTrack\Assemblies` with the following one:
https://github.com/AthenaSulisMinerva/CombatExtendedFastTrack/blob/c9c411e518cf2727e005f47fd63658cfa497c3b0/Assemblies/CombatExtended.dll?raw=true
4. Start a new savegame and see what happens
5. Replace the DLL with the following one from the original upstream CE mod:
https://github.com/CombatExtendedRWMod/CombatExtended/blob/287a2a58c6d1f57f2dca09adde4c0e92ba0b8dea/Assemblies/CombatExtended.dll?raw=true
6. Start a new savegame and see what happens

**Comment by Maeyanie** [comment]
> By these statements, we mean that we recognize XML patches written for the original upstream CE mod, which is what the vast majority of patches out there are. If a mod such as VEFS-CE wants to do unusual things with Harmony patching of C# assemblies, that's outside our guarantee.
Completely understandable, and the exact reason why I figured you still might not see it as a problem. But I wasn't sure, so I figured I'd leave it up to you guys to decide.

1 & 2: Running those (plus Fluffy's Mod Manager 2nd on the list) with -quicktest does load the map, with a number of errors, including the same loader exceptions. 
[output_log.txt](https://github.com/AthenaSulisMinerva/CombatExtendedFastTrack/files/4122574/output_log.txt)
Adding Humanoid Alien Races and O21's Elves to the end causes an error generating the map, so it seems the minimal set is still mangling Harmony.

3 & 4: After replacing that DLL, it loaded the map fine, with none of the assembly loading errors. There were a number of abstract class errors, probably due to a missing prereq. It doesn't list any that I can find, but almost everything needs Hugslib these days, so I wouldn't be surprised if that's the cause. Either way, not your problem.
[output_log.txt](https://github.com/AthenaSulisMinerva/CombatExtendedFastTrack/files/4122634/output_log.txt)

5 & 6: Identical to above. Map loaded fine, only abstract class errors.
[output_log.txt](https://github.com/AthenaSulisMinerva/CombatExtendedFastTrack/files/4122643/output_log.txt)

**Comment by Maeyanie** [comment]
I ran the test first with the same mod order and the 1.0.1.6 DLL from yesterday to make sure nothing changed from yesterday and it still worked. Fortunately, it did.

Replacing the DLL with the one from that link again caused it to fail to generate the map.
[output_log.txt](https://github.com/AthenaSulisMinerva/CombatExtendedFastTrack/files/4125907/output_log.txt)

**Comment by sumghai** [comment]
Alright, I think I have a good idea of what the problem is.

It seems that VEFS-CE is specifically looking for version 1.0.1.6 of the CE DLL, and will fail to work if it finds any other version. The strange thing is, nothing in the VEFS-CE codebase suggests that it is dependent on CE's assembly version number.

While we could, in theory, drop the version number for the DLL back down to 1.0.1.6, the problem is that users making bug reports (who insist on giving us log files, even though we have no use for them), keep quoting this old version number, confusing the hell out of us when we try to provide support because we kept getting mislead into thinking they are running older versions of CE:FT. So we up-rev'd the assembly version number, and that had the unfortunate coincidence of no longer being compatible with VEFS-CE.

Maintaining separate builds of the CE:FT DLL (one mainstream, one with the old version number) is not practical, because when VEFS-CE users encounter other problems, they'll keep coming to us reporting the wrong CE:FT version number from the logs over and over again.

Since VEFS-CE's current maintainer Andross says he doesn't do C#, we'll need to ask whoever wrote the VEFS-CE assembly (i.e. @xeonovadan) to update it so that it isn't so dependent on CE version numbers.

In other words, this is now a third-party mod issue.

**Comment by Maeyanie** [comment]
Alright. If they don't have a regular programmer, I'll take a look at their source and see what I can do... I have no experience with Harmony, but can't hurt.

**Comment by Maeyanie** [comment]
Well, all it took was recompiling it against your DLL. However, that prevents it working with regular CE. So it looks like if the version numbers are different, it'll be hard for any C#-based mod to support both versions simultaneously. I'll see if I can find a way.

**Comment by sumghai** [comment]
I think it's only a matter of time before the original upstream CE mod gets recompiled with an updated version number, if the original CE team doesn't want to get bug reports referencing the "older" version number.

**Comment by sumghai** [comment]
Sorry, fat fingered while on mobile.

**Comment by Maeyanie** [comment]
If you know a way to tell Visual Studio that, I'd love to hear it. The mod's code isn't specific at all, but .NET inherently links to the version included as a reference and no other version.

Loading a DLL manually is one work-around I found, but it requires knowing where to find the DLL to load, or it being in the search path, neither of which are true for a RimWorld mod.

**Comment by sumghai** [comment]
I'm still trying to loop @XeoNovaDan in, since he originally wrote the current code. I haven't heard back from him yet, though.

**Comment by Maeyanie** [comment]
Correction to the above, testing with a non-minimal modset shows including CombatExtended.dll with the assembly and relying on .NET redirection doesn't work.

Looking for a way to do it purely with reflection, but VS isn't much liking using variables as types.

**Comment by AthenaSulisMinerva** [comment]
@sumghai - Seeing as the original upstream CE mod assembly is still on 1.0.1.6, perhaps we should revert the internal assembly version number to 1.0.1.6, to placate VEFS-CE users?

If a user insists on giving us HugsLibs logs and quoting the outdated number, that's on them, not us. And if the upstream CE team also changes their assembly version, then that will be the time the ball falls back into VEFS-CE's court. Until then, let's try keep the users happy.

@Maeyanie - Are you absolutely certain that @sumghai can simply recompile the current assembly with the old version number, without removing the CE:FT-unique features?

**Comment by Maeyanie** [comment]
I can say with absolute certainty it's working for me, as I have a game running with that in the background as I'm typing this.
It's not only VEFS-CE which it broke, too. Initially I tried recompiling that with the 1.0.1.8 version and using it, but it also broke Misc. XPatches CE, which is closed-source so I couldn't easily recompile. I wouldn't be surprised if there are others out there too, since .NET seems to make it very difficult to do anything complicated with a DLL without compiling against it and creating a version-specific reference.

**Comment by sumghai** [comment]
I'll make the changes and publish this as FT-014 over the next hour or so.

@Maeyanie - CE:FT can never include Mod Manager or ModSync manifest files, because our version numbering system cannot be used with their standardized semantic version numbering format.

(Specifically, semantic version numbering limits modders to the `version.major.minor.patch` format, which the upstream CE mod has already taken for their current 1.0.1.8 release. CE:FT adds the -FT-xxx suffix to denote the extra patches we add for each update, which Mod Manager or ModSync will not recognize.

If we stick to semantic versioning and increment the number ourselves like, say 1.0.1.14, it would most likely cause confuse mod managers when the upstream CE mod actually reaches 1.0.1.14 themselves.)

**Comment by Maeyanie** [comment]
Thanks for reconsidering that. I never would have expected something as simple as a version number increment would have such far-reaching consequences which are so difficult to work around. This does explain why I see so many libraries permanently marked as version 1.0.0.0 however.
